### ISSUE TITLE
Accuracy of heap profile

### ISSUE THREAD

**Issue by jrudolph** [jvm-profiling-tools/async-profiler#51]
I've got a profile where 90%+ of allocations are object array allocations in one method which does not do any allocations in bytecode. What kind of inaccuracies are expected when doing heap profiling? Could it be allocated in inlined methods which are not shown?

**Comment by jrudolph** [comment]
In this particular case it seems to be a "sibling" method that is actually doing the allocations:

```
Counter: 9428816 (97.81%)
  [ 0] [Ljava/lang/Object;
  [ 1] akka.http.scaladsl.model.Uri$.create
  [ 2] akka.http.impl.model.parser.UriParser.parseHttpRequestTarget
  [ 3] akka.http.scaladsl.model.Uri$.parseHttpRequestTarget
  [ 4] akka.http.impl.engine.parsing.HttpRequestParser$$anon$1.parseRequestTarget
  [ 5] akka.http.impl.engine.parsing.HttpRequestParser$$anon$1.parseMessage
  [ 6] akka.http.impl.engine.parsing.HttpMessageParser$class.startNewMessage
  [ 7] akka.http.impl.engine.parsing.HttpRequestParser$$anon$1.startNewMessage
  [ 8] akka.http.impl.engine.parsing.HttpMessageParser$$anonfun$startNewMessage$1.apply
```

Using VisualVM I found out that it's not `create` but another call chain inside of `UriParser.parseHttpRequestTarget` which is actually doing the allocations.

**Comment by apangin** [comment]
This might be a bug of AsyncGetCallTrace. Does `-XX:+DebugNonSafepoints` make it better?
Do you have a small reproducer that you can share?

**Comment by retronym** [comment]
I've also run into seemingly impossible stack traces when CPU profiling the Scala compiler with flight recorder and async-profiler. For a while, I thought I was going crazy! It's comforting to know that the flaw might be elsewhere.

I've just trawled some open bugs on `AsyncGetCallTrace`. [JDK-8022893](https://bugs.openjdk.java.net/browse/JDK-8022893) is a good one with a reproduction and has [analysis of the root cause in the comments](https://bugs.openjdk.java.net/browse/JDK-8022893?focusedCommentId=13394602&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13394602)

**Comment by jrudolph** [comment]
@apangin in my case with `-XX:+DebugNonSafepoints` it is missing intermediate frames:

```
  [ 0] [Ljava/lang/Object;
  [ 1] akka.http.impl.engine.parsing.HttpRequestParser$$anon$1.parseRequestTarget
  [ 2] akka.http.impl.engine.parsing.HttpRequestParser$$anon$1.parseMessage
  [ 3] akka.http.impl.engine.parsing.HttpMessageParser$class.startNewMessage
  [ 4] akka.http.impl.engine.parsing.HttpRequestParser$$anon$1.startNewMessage
  [ 5] akka.http.impl.engine.parsing.HttpMessageParser$$anonfun$startNewMessage$1.apply
  [ 6] akka.http.impl.engine.parsing.HttpMessageParser$$anonfun$startNewMessage$1.apply
  [ 7] akka.http.impl.engine.parsing.HttpMessageParser$$anonfun$continue$2.apply
  [ 8] akka.http.impl.engine.parsing.HttpMessageParser$$anonfun$continue$2.apply
  [ 9] akka.http.impl.engine.parsing.HttpMessageParser$class.run$1
```

while it should have three more frames on top of `HttpRequestParser$$anon$1.parseRequestTarget`.

I don't have a small reproducer right now, I guess I could build a bigger one sharing a ~30MB jar file and instructions.

**Comment by apangin** [comment]
OK, looks like I can reproduce the problem. Will investigate this.

**Comment by apangin** [comment]
After some investigation I can confirm that it is AsyncGetCallTraces that works wrong.
If I replace AGCT with the standard JVM TI GetStackTrace call, everything looks good. Fortunately, it is safe to do so for allocation profiler (but not for CPU profiler).

I think the right fix would be to check if current thread has _last_Java_sp set, and, if yes, call JVM TI GetStackTrace(), otherwise call AsyncGetCallTrace.

**Comment by jrudolph** [comment]
Might make sense to compare results with the output generated by grav: https://epickrram.blogspot.de/2017/09/heap-allocation-flamegraphs.html

**Comment by apangin** [comment]
Re-implemented Java stack traces with JVM TI for allocation profiling. This is going to be accurate now.

**Comment by apangin** [comment]
Unfortunately, calling JVM TI from VM context may sometimes trigger assertion in VM code (#64), so reverted back to AsyncGetCallTrace for now. Better inaccurate than unsafe.
Will fix that properly in #66.

**Comment by apangin** [comment]
Re-implemented allocation stack walking once again. Now the stacks should be accurate without a risk of internal JVM deadlocks or assertion failures.
